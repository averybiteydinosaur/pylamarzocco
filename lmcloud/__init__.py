"""Import for ease of use."""

from .client_local import LaMarzoccoLocalClient
from .client_cloud import LaMarzoccoCloudClient
from .client_bluetooth import LaMarzoccoBluetoothClient
