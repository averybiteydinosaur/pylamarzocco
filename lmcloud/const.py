# URLs
TOKEN_URL = "https://cms.lamarzocco.io/oauth/v2/token"
CUSTOMER_URL = "https://cms.lamarzocco.io/api/customer"
GW_MACHINE_BASE_URL = "https://gw.lamarzocco.io/v1/home/machines"


# Key Names for Dictionaries
KEY = "key"
SERIAL_NUMBER = "serial_number"
MACHINE_NAME = "machine_name"
MODEL_NAME = "model_name"
STEAM_BOILER_NAME = "SteamBoiler"
COFFEE_BOILER_NAME = "CoffeeBoiler1"
BOILER_TARGET_TEMP = "boilerTargetTemperature"
BOILERS = "boilers"
PLUMBED_IN = "isPlumbedIn"
WEEKLY_SCHEDULING_CONFIG = "weeklySchedulingConfig"


# helper functions
# convert schedule return format to API expected input format
def convert_schedule(schedule):
    days = ["monday", "tuesday", "wednesday", "thursay", "friday", "saturday", "sunday"]
    schedule_conv = []
    for day in days: 
        d = schedule[day]
        schedule_conv.append({
            "day": str.upper(day),
            "enabled": d["enabled"],
            "on" : d["h_on"] + ":" + d["m_on"].zfill(2),
            "off": d["h_off"] + ":" + d["m_off"].zfill(2)
        })

        