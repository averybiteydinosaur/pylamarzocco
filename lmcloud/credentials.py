class Credentials:

    def __init__(self, client_id, client_secret, username, password):
        self._client_id = client_id
        self._client_secret = client_secret
        self._username = username
        self._password = password

    @property
    def client_id(self):
        return self._client_id

    @property
    def client_secret(self):
        return self._client_secret

    @property
    def username(self):
        return self._username

    @property
    def password(self):
        return self._password