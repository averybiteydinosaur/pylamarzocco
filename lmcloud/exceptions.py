class AuthFail(Exception):
    """Error to indicate there is invalid auth info."""

    def __init__(self, msg):
        super().__init__(msg)