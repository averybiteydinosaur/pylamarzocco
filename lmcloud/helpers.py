""" Helper functions for lmcloud. """

from typing import Any

from .const import LaMarzoccoBoilerType, PrebrewMode, WeekDay
from .lm_cloud import LaMarzoccoCloudSchedule
from .lm_machine import (LaMarzoccoBoiler, LaMarzoccoCoffeeStatistics,
                         LaMarzoccoPrebrewConfiguration, LaMarzoccoSchedule,
                         LaMarzoccoScheduleDay)


def schedule_to_request(schedule: LaMarzoccoSchedule) -> LaMarzoccoCloudSchedule:
    """convert schedule to API expected input format"""

    schedule_conv: LaMarzoccoCloudSchedule = {
        "enable": schedule.enabled,
        "days": [],
    }
    for day, schedule_day in schedule.days.items():
        # requests wants 00:00, response gives 24:00
        h_on = "00" if schedule_day.h_on == 24 else str(schedule_day.h_on).zfill(2)
        h_off = "00" if schedule_day.h_off == 24 else str(schedule_day.h_off).zfill(2)

        hh_mm_on = h_on + ":" + str(schedule_day.m_on).zfill(2)
        hh_mm_off = h_off + ":" + str(schedule_day.m_off).zfill(2)

        schedule_conv["days"].append(
            {
                "day": str.upper(day),
                "enable": schedule_day.enabled,
                "on": hh_mm_on,
                "off": hh_mm_off,
            }
        )
    return schedule_conv


def parse_schedule(schedule: dict[str, Any]) -> LaMarzoccoSchedule:
    """Parse schedule from API config object."""
    global_enable: bool = schedule["enabled"]
    days: dict[str, LaMarzoccoScheduleDay] = {}
    for weekday in WeekDay:
        day_settings = schedule[weekday]
        days[weekday] = LaMarzoccoScheduleDay(
            enabled=day_settings["enabled"],
            h_on=day_settings["h_on"],
            h_off=day_settings["h_off"],
            m_on=day_settings["m_on"],
            m_off=day_settings["m_off"],
        )
    return LaMarzoccoSchedule(enabled=global_enable, days=days)


def parse_boilers(
    boilers: list[dict[str, Any]]
) -> tuple[LaMarzoccoBoiler, LaMarzoccoBoiler]:
    """Parse boiler settings from API config object."""
    coffee_boiler = LaMarzoccoBoiler(False, 0, 0)
    steam_boiler = LaMarzoccoBoiler(False, 0, 0)
    for boiler in boilers:
        is_enabled = boiler["isEnabled"] == "Enabled"
        current_temp = boiler["current"]
        target_temp = boiler["target"]
        if boiler["id"] == LaMarzoccoBoilerType.COFFEE:
            coffee_boiler = LaMarzoccoBoiler(
                enabled=is_enabled,
                current_temperature=current_temp,
                target_temperature=target_temp,
            )
        elif boiler["id"] == LaMarzoccoBoilerType.STEAM:
            steam_boiler = LaMarzoccoBoiler(
                enabled=is_enabled,
                current_temperature=current_temp,
                target_temperature=target_temp,
            )

    return coffee_boiler, steam_boiler


def parse_preinfusion_settings(
    config: dict[str, Any]
) -> tuple[PrebrewMode, dict[int, LaMarzoccoPrebrewConfiguration]]:
    """Parse preinfusion settings from API config object."""
    parsed: dict[int, LaMarzoccoPrebrewConfiguration] = {}
    i = 1
    preinfusion_settings = config.get("preinfusionSettings", {})
    mode = PrebrewMode(preinfusion_settings.get("mode", "Disabled"))
    for group in preinfusion_settings.get("Group1", {}):
        parsed[i] = LaMarzoccoPrebrewConfiguration(
            on_time=group.get("preWetTime", 0),
            off_time=group.get("preWetHoldTime", 0),
        )

    return mode, parsed


def parse_coffee_doses(config: dict[str, Any]) -> tuple[dict[int, int], int | None]:
    """Parse doses from API config object."""
    parsed: dict[int, int] = {}
    i = 1
    group_capabilities = config.get("groupCapabilities", [])
    if len(group_capabilities) == 0:
        return parsed, None

    for dose in group_capabilities[0].get("doses", []):
        parsed[i] = dose.get("stopTarget", 0)
        i += 1
    dose_hot_water = config.get("teaDoses", {}).get("DoseA", {}).get("stopTarget", 0)
    return parsed, dose_hot_water


def parse_cloud_statistics(
    statistics: list[dict[str, Any]]
) -> LaMarzoccoCoffeeStatistics:
    """Parse statistics from API statistics object."""

    drink_stats: dict[int, int] = {}
    for stat in statistics:
        coffee_type: int = stat["coffeeType"]
        count: int = stat["count"]
        if 0 <= coffee_type < 4:
            drink_stats[coffee_type + 1] = count
        elif coffee_type == 4:
            continuous = count
        elif coffee_type == -1:
            total_flushing = count
    return LaMarzoccoCoffeeStatistics(drink_stats, continuous, total_flushing)


def parse_webhook_statistics(statistics: dict[str, Any]) -> LaMarzoccoCoffeeStatistics:
    """Parse statistics from webhook statistics object."""
    group = statistics["groups"][0]
    doses = group["doses"]
    drink_stats: dict[int, int] = {}
    for dose in doses:
        key: str = next(iter(dose))
        if key.startswith("Dose"):
            drink_stats[ord(key[-1]) - 64] = dose[key]
        elif key == "ContinuousDose":
            continuous = dose[key]
    total_flushing = group["clean"]
    return LaMarzoccoCoffeeStatistics(drink_stats, continuous, total_flushing)
