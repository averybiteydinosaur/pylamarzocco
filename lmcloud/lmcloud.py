from .const import *
from .credentials import Credentials
from .exceptions import *
from .lmlocalapi import LMLocalAPI
from authlib.integrations.base_client.errors import OAuthError
from authlib.integrations.httpx_client import AsyncOAuth2Client
from datetime import datetime
import asyncio, logging

_logger = logging.getLogger(__name__)

class LMCloud:  

    @property
    def client(self):
        # if self._client.token.is_expired:
        #     self._client.refresh_token(url=TOKEN_URL)
        return self._client

    @client.setter
    def client(self, value: AsyncOAuth2Client):
        self._client = value

    @property
    def machine_info(self):
        return self._machine_info

    @property
    def power_status(self):
        if self._lm_local_api:
            return self._lm_local_api.power_status
        else:
            self._sync_config_obj_task()
            if self._config["MACHINE_MODE"] == "BrewingMode":
                return "On"
            else:
                return "StandBy"

    def __init__(self):
        _logger.setLevel(logging.DEBUG)

    '''
    Initialize a cloud only client
    '''
    @classmethod
    async def create(cls, credentials: Credentials):
        self = cls()
        self.client = await self._connect(credentials)
        self._machine_info = await self._get_machine_info()
        self._gw_url_with_serial = GW_MACHINE_BASE_URL + "/" + self.machine_info[SERIAL_NUMBER]
        return self

    '''
    Also initialize a local API client
    '''
    @classmethod
    async def create_with_local_api(cls, credentials: Credentials, ip, port):
        self = cls()
        self.client = await self._connect(credentials)
        self._machine_info = await self._get_machine_info()
        self._lm_local_api = LMLocalAPI(local_ip=ip, local_port=port, local_bearer=self.machine_info[KEY])
        self._gw_url_with_serial = GW_MACHINE_BASE_URL + "/" + self.machine_info[SERIAL_NUMBER]
        return self
        
    '''
    Establish connection by building the OAuth client and requesting the token
    '''
    async def _connect(self, credentials: Credentials):
        client = AsyncOAuth2Client(
            client_id=credentials.client_id,
            client_secret=credentials.client_secret,
            token_endpoint=TOKEN_URL
        )

        headers = {
            "client_id": credentials.client_id,
            "client_secret": credentials.client_secret,
        }

        try:
            await client.fetch_token(
                url=TOKEN_URL,
                username=credentials.username,
                password=credentials.password,
                headers=headers,
            )
            return client
    
        except OAuthError as err:
            raise AuthFail("Authorization failure") from err

    async def _get_machine_info(self):
        response = await self.client.get(CUSTOMER_URL)
        if response.status_code == 200:
            machine_info = {}
            fleet = response.json()["data"]["fleet"][0]
            machine_info[KEY] = fleet["communicationKey"]
            machine_info[SERIAL_NUMBER] = fleet["machine"]["serialNumber"]
            machine_info[MACHINE_NAME] = fleet["name"]
            machine_info[MODEL_NAME] = fleet["machine"]["model"]["name"]
            return machine_info


    '''
    Turn power of machine on or off
    '''
    async def set_power(self, power_status):
        power_status = str.upper(power_status)
        if not power_status in ["ON", "STANDBY"]:
            msg = "Power status can only be on or standby"
            self._logger.debug()
            exit(1)
        else:
            data = {"status": power_status}
            url = f"{self._gw_url_with_serial}/status"
            response = await self.client.post(url, json=data)
            return response

    '''
    Turn Steamboiler on or off
    '''
    async def set_steam(self, steam_state:bool):
        if not type(steam_state) == bool:
            msg = "Steam state must be boolean"
            _logger.debug(msg)
            raise ValueError(msg)
        else:
            data = {"identifier": STEAM_BOILER_NAME, "state": steam_state}
            url = f"{self._gw_url_with_serial}/enable-boiler"
            response = await self.client.post(url, json=data)
            return response

    '''
    Set steamboiler temperature (in Celsius)
    '''
    async def set_steam_temp(self, temperature:int):
        if not type(temperature) == int:
            msg = "Steam temp must be integer"
            _logger.debug(msg)
            raise ValueError(msg)
        elif not temperature == 131 and not temperature == 128 and not temperature == 126:
            msg = "Steam temp must be one of 126, 128, 131 (°C)"
            _logger.debug(msg)
            raise ValueError(msg)
        else:
            data = { "identifier": STEAM_BOILER_NAME, "value": temperature}
            url = f"{self._gw_url_with_serial}/target-boiler"
            response = await self.client.post(url, json=data)
            return response

    '''
    Set coffee boiler temperature (in Celsius)
    '''
    async def set_coffee_temp(self, temperature:int):

        if temperature > 104 or temperature < 85:
            msg = "Coffee temp must be between 85 and 104 (°C)"
            _logger.debug(msg)
            raise ValueError(msg)
        else:
            temperature = round(temperature, 1)
            data = { "identifier": COFFEE_BOILER_NAME, "value": temperature}
            url = f"{self._gw_url_with_serial}/target-boiler"
            response = await self.client.post(url, json=data)
            return response

    '''
    Get configuration from cloud
    '''
    async def get_config(self):
        url = f"{self._gw_url_with_serial}/configuration"
        response = await self.client.get(url)
        if response.status_code == 200:
            return response.json()["data"]

    '''
    Load the config into a variable in this class
    '''
    async def _sync_config_obj(self):
        if self._config:
            # wait at least 10 seconds between config uüpdates to not flood the remote API
            if (datetime.now() - self._last_config_update).total_seconds() < 10:
                return
        self._config = await self.get_config()
        self._last_config_update = datetime.now()

    '''
    Call above function in a separate task
    '''
    def _sync_config_obj_task(self):
        loop = asyncio.get_running_loop()
        loop.create_task(self._sync_config_obj())

    '''
    Enable/Disable Pre-Brew or Pre-Infusion (mutually exclusive)
    '''
    async def _set_pre_brew_infusion(self, mode):
        if mode != "Disabled" and mode != "TypeB" and mode != "Enabled":
            msg = "Pre-Infusion/Pre-Brew can only be TypeB (PreInfusion), Enabled (Pre-Brew) or Disabled"
            _logger.debug(msg)
            raise ValueError(msg)
        else:
            url = f"{self._gw_url_with_serial}/enable-preinfusion"
            data = {"mode": mode}
            response = await self.client.post(url, json=data)
            return response

    '''
    Enable/Disable Pre-brew (Mode = Enabled)
    '''
    async def set_prebrew(self, mode):
        return await self._set_pre_brew_infusion(mode)

    '''
    Enable/Disable Pre-Infusion (Mode = TypeB)
    '''
    async def set_preinfusion(self, mode):
        return await self._set_pre_brew_infusion(mode)

    '''
    Set Pre-Brew details
    '''
    async def configure_prebrew(self, prebrewOnTime=5000, prebrewOffTime=5000):
        if type(prebrewOnTime) != int or type(prebrewOffTime) != int:
            msg = "Prebrew times must be in ms (integer)"
            _logger.debug(msg)
            raise ValueError(msg)
        else:
            if prebrewOnTime % 100 != 0 or prebrewOffTime % 100 != 0:
                msg = "Prebrew times must be multiple of 100"
                _logger.debug(msg)
                raise ValueError(msg)
            url = f"{self._gw_url_with_serial}/setting-preinfusion"
            data = {
                "button": "Continuous",
                "group": "Group1",
                "holdTimeMs": prebrewOffTime,
                "wetTimeMs": prebrewOnTime
            }

            response = await self.client.post(url, json=data)
            return response

    '''
    Enable or disable plumbin mode
    '''
    async def enable_plumbin(self, enable:bool):
        if not type(enable) == bool:
            msg = "Enable param must be boolean"
            _logger.debug(msg)
            raise ValueError(msg)
        else:
            data = {"enable": enable}
            url = f"{self._gw_url_with_serial}/enable-plumbin"
            response = await self.client.post(url, json=data)
            return response

    '''
    Set auto-on/off schedule
    days object:
    [
        {
            "day": "MONDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        },
        {
            "day": "TUESDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        },
        {
            "day": "WEDNESDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        },
        {
            "day": "THURSDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        },
        {
            "day": "FRIDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        },
        {
            "day": "SATURDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        },
        {
            "day": "SUNDAY",
            "enable": false,
            "off": "00:00",
            "on": "00:00"
        }
    ]
    '''
    async def configure_schedule(self, enable: bool, schedule: list):
        url = f"{self._gw_url_with_serial}/scheduling"
        data = {"enable": enable, "days": schedule}
        response = await self.client.post(url, json=data)
        return response