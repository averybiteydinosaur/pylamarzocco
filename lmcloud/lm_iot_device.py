"""Base class for all La Marzocco IoT devices."""

import asyncio
import logging
from abc import abstractmethod
from collections.abc import Callable, Coroutine
from functools import wraps
from typing import Any

from bleak import BleakError

from .const import LaMarzoccoFirmwareType
from .exceptions import (
    AuthFail,
    BluetoothConnectionFailed,
    ClientNotInitialized,
    RequestNotSuccessful,
)
from .helpers import parse_firmware
from .client_bluetooth import LaMarzoccoBluetoothClient
from .client_cloud import LaMarzoccoCloudClient
from .client_local import LaMarzoccoLocalClient
from .models import LaMarzoccoFirmware, LaMarzoccoStatistics

_LOGGER = logging.getLogger(__name__)


def cloud_only(func: Callable[..., Coroutine]) -> Callable:
    """Decorator to ensure that the cloud client is initialized"""

    @wraps(func)
    async def wrapper(*args: Any, **kwargs: dict) -> Any:
        machine: LaMarzoccoIoTDevice = args[0]
        if machine.cloud_client is None:
            raise ClientNotInitialized("Cloud client is not initialized.")
        return await func(*args, **kwargs)

    return wrapper


def cloud_and_bluetooth(func: Callable[..., Coroutine]) -> Callable:
    """Decorator to show that the command can be sent via Bluetooth or Cloud"""

    @wraps(func)
    async def wrapper(*args: Any, **kwargs: dict) -> Any:
        return await func(*args, **kwargs)

    return wrapper


class LaMarzoccoIoTDevice:
    """Base class for all La Marzocco devices"""

    cloud_client: LaMarzoccoCloudClient

    def __init__(
        self,
        model: str,
        serial_number: str,
        name: str,
        local_client: LaMarzoccoLocalClient | None = None,
        bluetooth_client: LaMarzoccoBluetoothClient | None = None,
    ) -> None:
        """Initializes a new LaMarzoccoMachine instance"""
        self.model: str = model
        self.serial_number: str = serial_number
        self.name: str = name
        self.turned_on = False
        self.doses: dict[int, float] = {}
        self.firmware: dict[LaMarzoccoFirmwareType, LaMarzoccoFirmware] = {}
        self.statistics: LaMarzoccoStatistics | None = None

        self._raw_config: dict[str, Any] | None = None
        self._bluetooth_client: LaMarzoccoBluetoothClient | None = bluetooth_client
        self._local_client: LaMarzoccoLocalClient | None = local_client
        self._update_lock = asyncio.Lock()

    def parse_config(self, raw_config: dict[str, Any]) -> None:
        """Parse the config object."""
        self.firmware = parse_firmware(raw_config["firmwareVersions"])

    @abstractmethod
    def parse_statistics(
        self, raw_statistics: list[dict[str, Any]]
    ) -> LaMarzoccoStatistics:
        """Parse the statistics object."""

    @property
    def bluetooth_connected(self) -> bool:
        """Return the connection status of the bluetooth client."""
        if self._bluetooth_client is None:
            return False
        return self._bluetooth_client.connected

    async def get_config(
        self,
        local_api_retry_delay: int = 3,
    ) -> None:
        """Update the machine status."""
        raw_config: dict[str, Any] = {}

        async with self._update_lock:
            # first, try to update locally
            if self._local_client is not None:
                try:
                    raw_config = await self._local_client.get_config()
                except AuthFail as exc:
                    _LOGGER.debug(
                        "Got 403 from local API, sending token request to cloud"
                    )
                    if self.cloud_client is None:
                        await self.cloud_client.token_command(self.serial_number)
                        await asyncio.sleep(local_api_retry_delay)
                        raw_config = await self._local_client.get_config()
                    else:
                        raise exc
                except RequestNotSuccessful as exc:
                    _LOGGER.warning(
                        "Could not connect to local API although initialized, "
                        + "falling back to cloud."
                    )
                    _LOGGER.debug(exc)
                    if self.cloud_client is None:
                        raise exc

            # if local update failed, try to update from cloud
            if self.cloud_client is not None and not raw_config:
                raw_config = await self.cloud_client.get_config(self.serial_number)

            self.parse_config(raw_config)

    @cloud_only
    async def get_statistics(self) -> None:
        """Update the statistics"""
        raw_statistics = await self.cloud_client.get_statistics(self.serial_number)
        self.statistics = self.parse_statistics(raw_statistics)

    @cloud_only
    async def get_firmware(self) -> None:
        """Update the firmware"""
        self.firmware = await self.cloud_client.get_firmware(self.serial_number)

    async def _bluetooth_command_with_cloud_fallback(
        self, command: str, **kwargs
    ) -> bool:
        """Send a command to the machine via Bluetooth, falling back to cloud if necessary."""

        # First, try with bluetooth
        if self._bluetooth_client is not None:
            func = getattr(self._bluetooth_client, command)
            try:
                await func(kwargs.copy().pop("serial_number"))
            except (BleakError, BluetoothConnectionFailed) as exc:
                msg = "Could not send command to bluetooth device, even though initalized."

                if self.cloud_client is None:
                    _LOGGER.error(
                        "%s Cloud client not initialized, cannot fallback. Full error %s",
                        msg,
                        exc,
                    )
                    return False

                _LOGGER.warning("%s Falling back to cloud", msg)
                _LOGGER.debug("Full error: %s", exc)
            else:
                return True

        # no bluetooth or failed, try with cloud
        if self.cloud_client is not None:
            func = getattr(self.cloud_client, command)
            if await func(**kwargs):
                return True
        return False

    def __str__(self) -> str:
        attributes = {
            key: value for key, value in vars(self).items() if not key.startswith("_")
        }
        return str(attributes)
