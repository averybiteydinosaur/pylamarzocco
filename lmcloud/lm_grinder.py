"""La Marzocco grinder module."""

from typing import Any

from .lm_iot_device import LaMarzoccoIoTDevice, LaMarzoccoStatistics
from .lm_client_local import LaMarzoccoLocalClient


class LaMarzoccoGrinder(LaMarzoccoIoTDevice):
    """Class for La Marzocco grinder"""

    def __init__(
        self,
        model: str,
        serial_number: str,
        name: str,
        local_client: LaMarzoccoLocalClient | None = None,
    ) -> None:
        """Initializes a new LaMarzoccoGrinder instance."""
        super().__init__(model, serial_number, name, local_client)
        self.bell_opened = False
        self.stand_by_time: int = 5
        self.led_enabled = False

    def parse_config(self, raw_config: dict[str, Any]) -> None:
        """Parse the config object."""
        super().parse_config(raw_config)
        self.led_enabled = raw_config["baristaLed"]
        self.bell_opened = raw_config["bellOpened"]
        self.stand_by_time = raw_config["standByTime"]
        for dose in raw_config["doses"]:
            self.doses[ord(dose["doseIndex"][-1]) - 64] = dose["target"]

    def parse_statistics(
        self, raw_statistics: list[dict[str, Any]]
    ) -> LaMarzoccoStatistics:
        """Parse the statistics object."""
        raise NotImplementedError
