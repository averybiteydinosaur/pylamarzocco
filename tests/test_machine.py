"""Test the LaMarzoccoMachine class."""

import pytest

from syrupy import SnapshotAssertion

from lmcloud.client_cloud import LaMarzoccoCloudClient
from lmcloud.lm_machine import LaMarzoccoMachine


@pytest.mark.asyncio
async def test_create(
    cloud_client: LaMarzoccoCloudClient,
    snapshot: SnapshotAssertion,
):
    """Test creation of a cloud client."""
    LaMarzoccoMachine.cloud_client = cloud_client

    machine = await LaMarzoccoMachine.create(
        model="GS3",
        serial_number="123456",
        name="MyMachine",
    )
    assert machine == snapshot
