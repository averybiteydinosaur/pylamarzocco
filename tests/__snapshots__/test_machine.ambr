# serializer version: 1
# name: test_create
  LaMarzoccoMachine(
    auto_on_off_enabled=False,
    auto_on_off_schedule=LaMarzoccoSchedule(enabled=True, days={<WeekDay.MONDAY: 'monday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0), <WeekDay.TUESDAY: 'tuesday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0), <WeekDay.WEDNESDAY: 'wednesday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0), <WeekDay.THURSDAY: 'thursday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0), <WeekDay.FRIDAY: 'friday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0), <WeekDay.SATURDAY: 'saturday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0), <WeekDay.SUNDAY: 'sunday'>: LaMarzoccoScheduleDay(enabled=True, h_on=6, h_off=16, m_on=0, m_off=0)}),
    bluetooth_connected=False,
    brew_active=False,
    brew_active_duration=0,
    cloud_client=LaMarzoccoCloudClient(
      password='password',
      username='username',
    ),
    coffee_boiler=LaMarzoccoBoiler(enabled=False, current_temperature=96.5, target_temperature=95),
    dose_hot_water=8,
    doses=dict({
      1: 135,
      2: 97,
      3: 108,
      4: 121,
    }),
    firmware=dict({
      <LaMarzoccoFirmwareType.GATEWAY: 'gateway'>: LaMarzoccoFirmware(current_version='v2.2-rc0', latest_version='v2.2-rc0'),
      <LaMarzoccoFirmwareType.MACHINE: 'machine'>: LaMarzoccoFirmware(current_version='1.11', latest_version='1.11'),
    }),
    model='GS3',
    name='MyMachine',
    plumbed_in=True,
    prebrew_configuration=dict({
      1: LaMarzoccoPrebrewConfiguration(on_time=2, off_time=2),
    }),
    prebrew_mode=<PrebrewMode.PREINFUSION: 'TypeB'>,
    serial_number='123456',
    statistics=LaMarzoccoCoffeeStatistics(drink_stats={1: 167, 2: 0, 3: 0, 4: 0}, continous=0, total_flushing=228),
    steam_boiler=LaMarzoccoBoiler(enabled=False, current_temperature=123.80000305175781, target_temperature=123.9000015258789),
    steam_level=1,
    turned_on=True,
    water_contact=True,
  )
# ---
